[iOS] Prefer video sources with hardware accelerated decoder.

jyavenard opened the pull request:
#### cb6bd9411341625cedfc80e421782d3e47c08fc3
<pre>
[iOS] Prefer video sources with hardware accelerated decoder.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267753">https://bugs.webkit.org/show_bug.cgi?id=267753</a>
<a href="https://rdar.apple.com/120679553">rdar://120679553</a>

Reviewed by Youenn Fablet.

The MediaPlayer::shouldCheckHardwareSupport got broken when the GPU process was enabled
and the code to ensure that hardware decoding was required to allow playback became
non-functional.

We correct this problem and make use of it with the WebM player so that if a VP8 or VP9 track
is present without a hardware decoder decoding will fail and we will fallback to the next source
alternative.

Adding tests

* LayoutTests/media/content/h264-aac-good.mp4: Added.
* LayoutTests/media/content/h264-aac-good.png: Added.
* LayoutTests/media/content/vp8-opus-bad.png: Added.
* LayoutTests/media/content/vp8-opus-bad.webm: Added.
* LayoutTests/media/ios/media-sources-selection-expected.html: Added.
* LayoutTests/media/ios/media-sources-selection.html: Added.
* LayoutTests/media/media-sources-selection-expected.html: Added.
* LayoutTests/media/media-sources-selection.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::setShouldCheckHardwareSupport):
(WebCore::MediaPlayer::shouldCheckHardwareSupport const): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setShouldCheckHardwareSupport):
(WebCore::MediaPlayerPrivateInterface::shouldCheckHardwareSupport const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::trackIsPlayable const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldCheckHardwareSupport const): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerShouldCheckHardwareSupport const):
(WebKit::RemoteMediaPlayerProxy::setShouldCheckHardwareSupport):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setShouldCheckHardwareSupport):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/273271@main">https://commits.webkit.org/273271@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3b47055e242a5ca709f2809de23199ab4eb1eca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36988 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37659 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31540 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36087 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16203 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30474 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35472 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11702 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31131 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10232 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31220 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38912 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31765 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31548 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36319 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10400 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8316 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34297 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12215 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30863 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8007 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10942 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11278 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -1211,6 +1210,12 @@ void RemoteMediaPlayerProxy::playerContentBoxRectChanged(const WebCore::LayoutRe
+void RemoteMediaPlayerProxy::setShouldCheckHardwareSupport(bool value)
+{
+    m_player->setShouldCheckHardwareSupport(value);
+    m_shouldCheckHardwareSupport = value;
```
This is a bit odd.
It seems to me this would be more easily given as a parameter to MediaPlayer::load CRABS and MSE methods and passed down to MediaPlayerPrivate at that time.

That would allow to remove these getters and setters, but might make a bigger patch.
Could that be done as a follow-up?

jyavenard replied:
The mediaplayer construction in the GPU process is rather cumbersome

In the content process the MediaPlayer creates a MediaPlayerPrivate which here will be a MediaPlayerPrivateRemote. So we have to pass the information to the private player. 

The remote player however, talks to the GPU’s MediaPlayerProxy which is a MediaPlayerClient which will create a new MediaPlayerPrivate. 

So while in the CP, the private need the information, in the GPU it’s the MediaPlayerClient and MediaPlayer. 

I re-used the architecture in place, using setters felt like a less intrusive change over what was existing compare to passing parameters.

youennf commented on LayoutTests/media/ios/media-sources-selection.html:
```diff
@@ -0,0 +1,35 @@
+<html>
+<meta name="fuzzy" content="maxDifference=0-103; totalPixels=0-329652" /><head>
```
Can we run this test for macOS as well?
Can we add a similar test for macOS + CRABS?

jyavenard replied:
The change only applies to iOS

jyavenard replied:
I added test for non-iOS ; to show that the first image is selected

jyavenard committed:
[iOS] Prefer video sources with hardware accelerated decoder.
https://bugs.webkit.org/show_bug.cgi?id=267753
rdar://120679553

Reviewed by Youenn Fablet.

The MediaPlayer::shouldCheckHardwareSupport got broken when the GPU process was enabled
and the code to ensure that hardware decoding was required to allow playback became
non-functional.

We correct this problem and make use of it with the WebM player so that if a VP8 or VP9 track
is present without a hardware decoder decoding will fail and we will fallback to the next source
alternative.

Adding tests

* LayoutTests/media/content/h264-aac-good.mp4: Added.
* LayoutTests/media/content/h264-aac-good.png: Added.
* LayoutTests/media/content/vp8-opus-bad.png: Added.
* LayoutTests/media/content/vp8-opus-bad.webm: Added.
* LayoutTests/media/ios/media-sources-selection-expected.html: Added.
* LayoutTests/media/ios/media-sources-selection.html: Added.
* LayoutTests/media/media-sources-selection-expected.html: Added.
* LayoutTests/media/media-sources-selection.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::setShouldCheckHardwareSupport):
(WebCore::MediaPlayer::shouldCheckHardwareSupport const): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setShouldCheckHardwareSupport):
(WebCore::MediaPlayerPrivateInterface::shouldCheckHardwareSupport const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::trackIsPlayable const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldCheckHardwareSupport const): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerShouldCheckHardwareSupport const):
(WebKit::RemoteMediaPlayerProxy::setShouldCheckHardwareSupport):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setShouldCheckHardwareSupport):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: https://commits.webkit.org/273271@main
